src/python/pants/backend/kotlin/dependency_inference/kotlin_parser.py

```diff
@@ -59,14 +61,64 @@ def from_json_dict(cls, d: dict) -> KotlinImport:
 
 @dataclass(frozen=True)
 class KotlinSourceDependencyAnalysis:
+    package: str
     imports: frozenset[KotlinImport]
     named_declarations: frozenset[str]
+    consumed_symbols_by_scope: FrozenDict[str, frozenset[str]]
+    scopes: frozenset[str]
+
+    def fully_qualified_consumed_symbols(self) -> Iterator[str]:
```

Comment:
This code is fiddly enough in Scala that it would be good to see how much can be reused (perhaps with a common base class and smaller reusable functions). It looks like `kotlin` doesn't support imports anywhere other than at the top of a file, which makes for annoying difference.

Reply:
The only difference in the algorithm is the for loop over parent scopes where for Kotlin it only loops if the "parent scope" is the file's package scope. So definitely room for refactoring. I am going to land this and then refactor both Kotlin and Scala.